teach the tool about copyright lines

I keep seeing warnings of the form:

```
/home/mvdan/src/sh/shell/doc.go:1:1: "Martí" is misspelled in comment
/home/mvdan/src/sh/shell/doc.go:1:1: "mvdan" is misspelled in comment
	Copyright (c) 2017, Daniel Martí <mvdan@mvdan.cc>
```

Presumably I shouldn't need to add my username or full name to a `.words` file in every project :) It should be fairly easy to discard a comment if it looks like a copyright line. Another route would be to ignore `Full Name <email@domain.com>`.

--- a/README.md
+++ b/README.md
@@ -112,6 +112,7 @@ TOML format `.gospel.conf` file. A number of options are provided:
 - `ignore_upper` — whether to ignore words that are all uppercase or their plurals.
 - `ignore_single` — whether to ignore single rune words.
 - `ignore_numbers` — whether to ignore number literals.
+- `read_licenses` — whether to ignore words found in license files.
 - `mask_flags` — whether words that could be command-line flags should be removed prior to checking.
 - `mask_urls` — whether URLs should be removed prior to checking.
 - `check_urls` — whether the HTTP/HTTPS reachability of URLs should be checked.
@@ -138,6 +139,7 @@ check_embedded = false
 ignore_upper = true
 ignore_single = true
 ignore_numbers = true
+read_licenses = true
 mask_flags = false
 mask_urls = true
 check_urls = false

--- a/README.tmpl.md
+++ b/README.tmpl.md
@@ -105,6 +105,7 @@ TOML format `.gospel.conf` file. A number of options are provided:
 - `ignore_upper` — whether to ignore words that are all uppercase or their plurals.
 - `ignore_single` — whether to ignore single rune words.
 - `ignore_numbers` — whether to ignore number literals.
+- `read_licenses` — whether to ignore words found in license files.
 - `mask_flags` — whether words that could be command-line flags should be removed prior to checking.
 - `mask_urls` — whether URLs should be removed prior to checking.
 - `check_urls` — whether the HTTP/HTTPS reachability of URLs should be checked.

--- a/config.go
+++ b/config.go
@@ -35,6 +35,7 @@ type config struct {
 	IgnoreUpper     bool          `toml:"ignore_upper"`   // ignore words that are all uppercase.
 	IgnoreSingle    bool          `toml:"ignore_single"`  // ignore words that are a single rune.
 	IgnoreNumbers   bool          `toml:"ignore_numbers"` // ignore Go syntax number literals.
+	ReadLicenses    bool          `toml:"read_licenses"`  // ignore all words found in license files.
 	MaskFlags       bool          `toml:"mask_flags"`     // ignore words with a leading dash.
 	MaskURLs        bool          `toml:"mask_urls"`      // mask URLs before checking.
 	CheckURLs       bool          `toml:"check_urls"`     // check URLs point to reachable targets.
@@ -66,6 +67,7 @@ var defaults = config{
 	IgnoreUpper:     true,
 	IgnoreSingle:    true,
 	IgnoreNumbers:   true,
+	ReadLicenses:    true,
 	MaskFlags:       false,
 	MaskURLs:        true,
 	CheckURLs:       false,

--- a/dictionary.go
+++ b/dictionary.go
@@ -142,6 +142,13 @@ func newDictionary(pkgs []*packages.Package, cfg config) (*dictionary, error) {
 	}
 	ook.urls = nil
 
+	if cfg.ReadLicenses {
+		const licenseThreshold = 75 // Threshold for matching a license.
+		for r := range d.roots {
+			readLicenses(d.Spell, r, licenseThreshold)
+		}
+	}
+
 	if cfg.IgnoreIdents {
 		err = addIdentifiers(d.Spell, pkgs, make(map[string]bool))
 		if err != nil {

--- a/go.mod
+++ b/go.mod
@@ -5,6 +5,7 @@ go 1.17
 require (
 	github.com/BurntSushi/toml v1.0.0
 	github.com/google/go-cmp v0.5.7
+	github.com/google/licensecheck v0.3.1
 	github.com/kortschak/camel v0.0.0-20220208065757-0665e01e7dba
 	github.com/kortschak/ct v0.0.0-20140325011614-7d86dffe6951
 	github.com/kortschak/hunspell v0.0.0-20220305030544-5d8374a03860

--- a/licenses.go
+++ b/licenses.go
@@ -1,0 +1,98 @@
+// Copyright ©2022 Dan Kortschak. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"bufio"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+	"unicode"
+
+	"github.com/google/licensecheck"
+	"github.com/kortschak/hunspell"
+)
+
+// readLicenses adds words from licenses under root that satisfy the licensecheck
+// threshold provided.
+func readLicenses(spelling *hunspell.Spell, root string, thresh float64) error {
+	texts, err := licenses(root, thresh)
+	if err != nil {
+		return err
+	}
+	for _, text := range texts {
+		sc := bufio.NewScanner(strings.NewReader(text))
+		var w words // Use our word scanner to retain parity.
+		sc.Split(w.ScanWords)
+		for sc.Scan() {
+			w := quietly(sc.Text())
+			if spelling.IsCorrect(w) {
+				continue
+			}
+			spelling.Add(w)
+		}
+	}
+	return nil
+}
+
+// quietly returns the provided string lower cased if it is all upper case.
+func quietly(s string) string {
+	for _, r := range s {
+		if !unicode.IsUpper(r) {
+			return s
+		}
+	}
+	return strings.ToLower(s)
+}
+
+// licenses returns the text of all files matching licenses using
+// licensecheck.Scan with at least a thresh match.
+func licenses(root string, thresh float64) ([]string, error) {
+	maybeLicense := make(map[string]bool)
+	for _, c := range candidates {
+		maybeLicense[strings.ToLower(c)] = true
+	}
+
+	var texts []string
+	err := filepath.WalkDir(root, func(path string, info fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		name := strings.TrimSuffix(info.Name(), filepath.Ext(info.Name()))
+		if !maybeLicense[strings.ToLower(name)] {
+			return nil
+		}
+		b, err := os.ReadFile(info.Name())
+		if err != nil {
+			return err
+		}
+		if licensecheck.Scan(b).Percent >= thresh {
+			texts = append(texts, string(b))
+		}
+		return nil
+	})
+	return texts, err
+}
+
+var candidates = []string{
+	"COPYING",
+	"LICENCE",
+	"LICENSE",
+	"LICENSE-2.0",
+	"LICENCE-2.0",
+	"LICENSE-APACHE",
+	"LICENCE-APACHE",
+	"LICENSE-APACHE-2.0",
+	"LICENCE-APACHE-2.0",
+	"LICENSE-MIT",
+	"LICENCE-MIT",
+	"MIT-LICENSE",
+	"MIT-LICENCE",
+	"MIT_LICENSE",
+	"MIT_LICENCE",
+	"UNLICENSE",
+	"UNLICENCE",
+}

--- a/main.go
+++ b/main.go
@@ -39,6 +39,7 @@ func gospel() (status int) {
 	flag.BoolVar(&config.IgnoreUpper, "ignore-upper", config.IgnoreUpper, "ignore all-uppercase words")
 	flag.BoolVar(&config.IgnoreSingle, "ignore-single", config.IgnoreSingle, "ignore single letter words")
 	flag.BoolVar(&config.IgnoreNumbers, "ignore-numbers", config.IgnoreNumbers, "ignore Go syntax number literals")
+	flag.BoolVar(&config.ReadLicenses, "read-licenses", config.ReadLicenses, "ignore words found in license files")
 	flag.BoolVar(&config.MaskFlags, "mask-flags", config.MaskFlags, "ignore words with a leading dash")
 	flag.BoolVar(&config.MaskURLs, "mask-urls", config.MaskURLs, "mask URLs in text")
 	flag.BoolVar(&config.CheckURLs, "check-urls", config.CheckURLs, "check URLs in text with HEAD request")

--- a/testdata/license_text.txt
+++ b/testdata/license_text.txt
@@ -1,0 +1,48 @@
+# Show license text suppresses detection when requested.
+
+! gospel -read-licenses=false
+! stderr .
+cmp stdout expected_output
+
+gospel -read-licenses=true
+! stdout .
+! stderr .
+
+-- go.mod --
+module dummy
+-- main.go --
+// Copyright ©2022 Dan Kortschak. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+func main() {
+}
+-- LICENSE.txt --
+Copyright ©2022 Dan Kortschak. All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    * The name of the author may not be used to endorse or promote products
+      derived from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+-- expected_output --
+main.go:1:25: "Kortschak" is misspelled in comment
+	// Copyright ©2022 Dan [31;1;3mKortschak[0m. All rights reserved.
+	// Use of this source code is governed by a BSD-style

--- a/testdata/write_config.txt
+++ b/testdata/write_config.txt
@@ -19,6 +19,7 @@ check_embedded = false
 ignore_upper = true
 ignore_single = true
 ignore_numbers = true
+read_licenses = true
 mask_flags = false
 mask_urls = true
 check_urls = false

